### ISSUE TITLE
Github automation for running builds, style guide checks and unit tests

### ISSUE THREAD

**Issue by speckdavid** [prost-planner/prost#104]
We want to use github actions to perform automated checks, such as:
- Builds
- Style guide checks
- Unit tests
- Run small instance for performance checks

**Comment by speckdavid** [comment]
I have added a github action file that performs a release build and a debug build (including unit tests). These github actions are currently executed for each push and pull request for the Issue 104 branch. However, we can specify for which branch we want to do this.

Here you can see the result: https://github.com/thomaskeller79/prost-dev/runs/742050911

A few remarks:
- I want to add an additional action that checks the style guide and returns the files that are not correctly formatted (clang format and black). Perhaps the best option is to write a small Python script that does this and returns the desired exit code.
- Currently, a commit is successful even if the Planner build fails. I assume that the current build.py script always terminates with an exit code of 0. We should probably change this if the build fails.

**Comment by speckdavid** [comment]
I have now added a style check script and github action. The style check is currently failing, because not all files are formatted properly. 

https://github.com/thomaskeller79/prost-dev/runs/742275767

MAIL:

Run failed for issue-104 (f25b6f6)

Repository: thomaskeller79/prost-dev
Workflow: Prost Github Actions
Duration: 3 minutes and 1.0 second
Finished: 2020-06-05 12:53:49 UTC

View results
Jobs:

    build succeeded (0 annotations)
    style failed (1 annotation)

**Comment by thomaskeller79** [comment]
This looks already great and it seems quite simple to set up. How does the system decide if a "run" fails or succeeds? I assume it is currently based on the processes' return value. Is it also possible to parse the output of a process (without a wrapper script that checks this), e.g., to determine if some process like the parser or a planning step with a fixed number of trials takes significantly longer than before?

**Comment by speckdavid** [comment]
> This looks already great and it seems quite simple to set up. How does the system decide if a "run" fails or succeeds? I assume it is currently based on the processes' return value.

Exactly, an exit code of 0 indicates that the action was successful and any other exit code indicates that the action failed.

> Is it also possible to parse the output of a process (without a wrapper script that checks this), e.g., to determine if some process like the parser or a planning step with a fixed number of trials takes significantly longer than before?

That's an interesting question. I think it is possible. At least you can make use of the output (https://help.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#outputs). But it looks a bit more complicated.

**Comment by speckdavid** [comment]
I have now included badges for build, test and style in the readme file of issue 104. For this it was necessary to make the readme file a markdown readme. I used the change to clean up the readme file. 

One problem with the badges is that they point to a specific branch and there is no way to use a placeholder of a branch name which is automatically replaced by the current branch. It seems that this feature is in high demand, but github has not implemented it yet (https://github.com/github/markup/issues/913).

I think the best option in the long run is to simply link the badges to the master branch. The badges would then always show the status of the master branch. The status of the current branch can be seen in the commit section of the current branch.

**Comment by speckdavid** [comment]
Currently we are waiting with the integration of github actions until some ongoing issues and the corresponding code changes are completed. This is mainly because we want to format all files according to our style guide, which can lead to big diffs.

**Comment by geisserf** [comment]
Additionally we want to update clang-format rules and the underlying clang-format version in this issue.
